### DISCUSSION TITLE
Testing extension, ability to show an error in the UI

### DISCUSSION THREAD

**Discussion by shadow-cs** [microsoft/vscode-discussions#722] [Extension Development QnA] [answered]
Hi, how are these UI elements supposed to work:

## `Peek error` 

When I right-click on the test glyph in the gutter and select it, nothing happens. Is there some special treatment I have to do in the extension to show the error?

I can see the error only in the `TEST RESULTS` tab (but navigating there and finding the right test may be difficult, if there are a lot of failed tests). Also, it would be nice if I could navigate from `Test explorer` to test output (I couldn't find an icon or menu item to do that).

## Test output

The [docs](https://code.visualstudio.com/api/extension-guides/testing#test-output) say:

> This output can be displayed in a terminal using the **Test: Show Output** and through various buttons in the UI, such as the terminal icon in the Test Explorer view.

But again I was only able to find them using the `TEST RESULTS`, I couldn't find any UI element described above.

Is there something wrong with my extensions that needs fixing so I can see the UI elements?

Any guidance will be appreciated, thanks.

**Comment by isidorn** [comment]
I suggest to start with the Test Provider sample extension https://github.com/microsoft/vscode-extension-samples/tree/main/test-provider-sample

If you hit some issues you can provide the exact error here.

**Reply by shadow-cs** [reply]
I did, and I don't have any issues but somewhat unexpected UX behavior.

To be more exact by not having issues I mean:

- I'm able to discover and run tests
- The tests and their outcome are appropriately displayed in Test explorer
- Test results show the detail of a failure and additional test output

So I was just wondering what needs to be done in order for the Peek output command to work (is there some user docs so I can learn more maybe?) and also what UI elements the docs are referring to in regard to the Test output?

**Reply by isidorn** [reply]
Peek test output should just work. When you click on the failed test it should show the reason why.

Here are the testing docs for Python and for Java. I think those should help as examples
https://code.visualstudio.com/docs/python/testing
https://code.visualstudio.com/docs/java/java-testing

**Reply by shadow-cs** [reply]
> Peek test output should just work.

I hoped so, but unfortunately, it doesn't. Going through the links you posted @isidorn and VSC Settings, I'm just wondering whether setting tests `range` after the test run in the extension can be an issue? I'm doing this lookup later (in the background) to not slow the test run.

**Reply by theReynald** [reply]
So we have more context, can you share a screenshot of what you're seeing and another with what you expect (just modify the image)?

**Reply by shadow-cs** [reply]
@theReynald thanks for the reply.

Regarding `Show output` I did retry and the problem was that I thought it is working differently, thus I created a wrong mental model around this. Console icon in the `Test explorer` shows the output for the last test run (not the focused test in the panel).

However `Peek error` (right-clicking on the gutter of a failed test) does nothing. I can see no error and no UI changes. Not sure what the purpose of this context menu is though.

![image](https://github.com/microsoft/vscode-discussions/assets/4301974/45d0f48e-e738-42cc-9771-bee79e240a69)

I'd expect it'll show the `Test results` with the error highlighted. I'd be the happiest if just hovering on the Failure icon in the gutter could provide me with the failure context (inside a hint). In my case the first item in the `Test results` tree for the failed test, for example: `expected: <Foo> but was: <Bar>`. Instead of (or in addition to):

![image](https://github.com/microsoft/vscode-discussions/assets/4301974/dbe45dd2-82e7-4871-9ea5-19d5843b3d0e)

The same applies to hints in the `Test explorer`, providing some feedback without further navigation would be most helpful.

And again, I'm not saying my extension does everything "by the book". I did try my best with the docs and samples, I didn't go through the source of more complex testing extensions though.

**Comment by theReynald** [comment]
Hi @shadow-cs ...

In VS Code, "Peek error" is a feature that allows you to quickly view the details of an error directly in the context of your code without having to navigate away from your current location. When you have test results or compiler errors that VS Code can detect, right-clicking on the gutter (the space to the left of the line numbers) where a failed test or error is indicated and selecting "Peek error" will open an inline window, or "peek", that shows the error message and possibly a stack trace or other relevant information.

You'd click on "Peek error" when you want to:

Quickly view the error or failed test details.
Get a better understanding of what the problem is without losing your place in the code.
Avoid the disruption of switching between files or panels to find the error details.
For example, suppose you are writing a suite of unit tests and one of them fails. Next to the line number of the failing test, you might see a red squiggle or an icon indicating an issue. By right-clicking on the gutter next to this indicator and choosing "Peek error," a small window will open up beneath the line of code with the failing test's error message. This allows you to see what went wrong immediately, like an assertion failure message stating expected values did not match the actual values.

If "Peek error" does nothing when you try to use it, it could be due to several reasons:

- The feature might not be implemented for the specific language or framework you are working with.
- There might be a problem with your VS Code setup or a conflict with another extension.
- The test framework might not be properly integrated with VS Code's testing interface.

In such cases, you might need to look into the VS Code documentation for your specific setup, check if the language or test framework you're using supports this feature, or potentially debug your VS Code installation and extensions.

**Reply by shadow-cs** [reply]
Hi @theReynald and thanks again.

From what you said, `Peek error` does precisely what I assume it should do 😉 👍.

I probably didn't implement something to get `Peek error` to work in my extension. But I don't know what the "something" is, can you point me towards some documentation or an example please?

**Comment by theReynald** [comment]
Sure. You can refer to the official documentation on the Testing API:

- [Testing API](https://code.visualstudio.com/api/extension-guides/testing): This guide covers how to implement testing within VS Code extensions, including how to handle test runs, report test results, and deal with errors.

The section on running tests and handling test results will be particularly relevant for the "Peek error" feature, as it relies on the extension properly reporting errors for each test case using the `TestRun` API.

**Reply by shadow-cs** [reply]
@theReynald thanks, but I did read this guide and I think I do everything described there, so maybe I'm missing some bit which makes `Peek error` inoperative. 

I do see my test runs in `Test results`, the method executions are correctly reported (along with the result of the particular test) and I also see the test output there (after selecting the test method).

So any more insight into how to fix this problem would be greatly appreciated.

**Comment by burkeholland** [comment]
Pinged @connor4312 here since this is something that should be “just working” as @isidorn mentioned.

**Comment by connor4312** [comment, marked as answer]
Thanks for the discussion, I made an issue to fix the case I think you're hitting: https://github.com/microsoft/vscode/issues/199552

**Reply by shadow-cs** [reply]
@connor4312 yes that is the case. Is there some way I can test this before release, or should I just wait for it to land?

You're talking about `TestMessage.location` right? I'm indeed not setting this. I, unfortunately, missed this entirely in both [Testing API](https://code.visualstudio.com/api/extension-guides/testing) (it is just in a comment) and the sample extension (`diff` method is used there). Giving the `TestMessage` class a small chapter would make it much clearer IMO.

**Reply by shadow-cs** [reply]
I tried insiders build, but I still see `Peek error` and it does nothing.

**Reply by connor4312** [reply]
Can you provide instructions so I can try to reproduce the problem you're seeing?

**Reply by shadow-cs** [reply]
@connor4312 I find it difficult to explain (my extension is using sort-of a "Publish only" approach together with lazy loading of some details, maybe this is non-standard approach). So I created a private [repo](https://github.com/shadow-cs/testinsight-vsc/) for my extension and invited you there (I'm not quite ready to make it public yet, I don't feel the quality is good enough) so you can take a look and we can help each other out 😉. Thanks.

**Comment by burkeholland** [comment]
Closing this in favor of @connor4312 issue.
